### DISCUSSION TITLE
the-dangers-and-gotchas-of-using-scoped-services-when-configuring-options-with-options-builder/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#83] [Announcements]
# the-dangers-and-gotchas-of-using-scoped-services-when-configuring-options-with-options-builder/

In this post I look at the problems you can hit with strongly-typed settings when you inject Scoped services into OptionsBuilder, and how to avoid them

https://andrewlock.net/the-dangers-and-gotchas-of-using-scoped-services-when-configuring-options-with-options-builder/

**Comment by jnyrup** [comment]
Always a joy to read you writings.
It's not only "here's a shorter way to do it" but also the differences in internal and observable behavior and their pitfalls.

... and now for the typo.
For the "Named options" section:

```cs
app.MapGet("/", (IOptionsSnapshot<MySettings> settings) => new {
    aliceSettings = settings.Get("Alice"), // get the Alice settings
    bobSettings = settings.Get("Bob"), // get the Bob settings
    mySettings = settings.Value, // get the default, unnamed settings
});
app.Run();
```

**Reply by andrewlock** [reply]
Hah, that's what I get for not proof reading 😅 Thanks!

**Reply by skovsende** [reply]
That's what we're here for! ;)

**Comment by jernejg** [comment]
>In contrast, if you inject IOptionsSnapshot<MySettings> into the endpoint, MySettings is re-bound every request, and OptionsBuilder<T>.Configure() is invoked on every request. That gives you a new value every time.

This would also mean we would have a blocking operation on every request right? There is not `async` context here and we are talking about configuring our options objects using complex operations, like making DB calls which should be awaited.

**Reply by andrewlock** [reply]
Yep, you can't use `async` in these code paths

**Comment by ThomasWetzel** [comment]
`public class ConfigureMySettingsOptions : IConfigureOptions<MySettings>
{
    // Can use standard DI here
    private readonly CalculatorService _calculator;
    public ConfigureMySettingsOptions(CalculatorService calculator)
    {
        _calculator = calculator;
    }

    public void Configure(MySettings options)
    {
        options.MyValue = _someService.DoComplexCalcaultion();
    }
}`

Should perhaps be

`public class ConfigureMySettingsOptions : IConfigureOptions<MySettings>
{
    // Can use standard DI here
    private readonly CalculatorService _calculator;
    public ConfigureMySettingsOptions(CalculatorService calculator)
    {
        _calculator = calculator;
    }

    public void Configure(MySettings options)
    {
        options.MyValue = _calculator.DoComplexCalculation();
    }
}`

**Comment by nhwilly** [comment]
Wow. Great explanation. Thank you.

**Comment by TallBaldGeek** [comment]
Thanks for the good read! 

I would suggest extracting the text from that exception screenshot and putting it in the post body itself. It might help future searchers who get a similar exception understand what's going on.

**Comment by charring** [comment]
According to this bug report, BindConfiguration() is broken.

https://github.com/dotnet/runtime/issues/85665

**Comment by lizanx** [comment]
> Behind the scenes the OptionsBuilder.Configure<T> registers an IConfigureOptions<T> object as a **transient** service.   

Maybe a typo here? "transient" should be "singleton" by my understanding.

Still, thanks for the great post.

**Comment by a14907** [comment]
To be honest, Microsoft makes the configuration related api a bit complicated. I usually use dependency injection directly to inject the required configuration (usually a singleton)!
